[CK_TILE] Stream-K Tile Partitioner Reboot

ecamartins opened the pull request:
## Proposed changes

CK Tile's current Stream-K Tile Partitioner does not align with the original Stream-K paper. Thus, this PR, done in collaboration with @arai713, introduces a new Stream-K Tile Partitioner that better aligns with the Stream-K paper. There are also other AMD teams working on Stream-K; thus, some of our new design considerations are more in line with what other teams are doing. The differences between the old and new CK Tile Stream-K Tile Partitioners are as follows:
- The grid size calculation now falls into 3 cases: (1) Stream-K only, (2) Data-parallel (DP) only, (3) DP + 2 Tile Stream-K
- When we perform a hybrid of DP and Stream-K, we perform the DP section before the Stream-K section
- We add a tile partitioner that supports a DP sections that are fully persistent or non-persistent (i.e. 1 C macro tile per work group).

We introduce a Stream-K Tile Partitioner base and derived class. There are two different template specializations for the derived class: one where persistent is true and another where persistent is false. Vanilla Stream-K itself is persistent; that said, the DP portion could be persistent or not. Thus, the two specializations for the derived classes handle the differences between when the DP portion of the algorithm is persistent or non-persistent. 

We have also added unit tests for our new tile partitioner.

Note: We will remove the old CK Tile Stream-K implementation and other associated artifacts in a future PR.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [x] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [ ] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

arai713 committed:
Addition of the derived structs for the new Stream-K TilePartitioner

There are 2 derived structs based on whether Stream-K is persistent or not.
If it's persistent that means that both the data parallel and Stream-K sections
are data parallel. If it's non-persistent that means that only the
Stream-K section is persistent, while the data parallel section will have
separate workgroups allocated for it. Both structs will have a template
argument for Persistent.

The 2 derived classes will inherit common variables and functions from the
Stream-K TilePartitioner base class. There are additional variables for the
differing data parallel sections that will be added to each derived class,
that are in charge of the indexing/bookkeeping for the data parallel sections.
The only additional function that will differ between the 2 structs is GridSize(),
as the non-persistent will allocate extra workgroups for data parallel.

Unit tests for the derived structs are included.

ecamartins committed:
Stream-K Tile Partitioner Base Class with Tests

To better align with the original Stream-K paper, this change implements
a new Stream-K tile partitioner base class. This class will handle the
Stream-K setup that is common to both a persistent and non-persistent DP
section. A later change will implement derived classes to handle the
differences between persistent and non-persistent DP.

This change also includes unit tests for the base tile partitioner.

cgmillette commented on include/ck_tile/ops/gemm/kernel/gemm_tile_partitioner.hpp:
```diff
@@ -10,6 +10,8 @@
 #include "ck_tile/core.hpp"
 #include "ck_tile/ops/common.hpp"
+#include <format>
+#include <iostream>
```
Are these needed? No format / iostream utilities used in this file

ecamartins replied:
Good catch. I've removed these in [a4689cb](https://github.com/ROCm/composable_kernel/pull/3018/commits/a4689cb90608d8985e2df8ee3cf5bc889a9f137f).

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm_tile_partitioner.hpp:
```diff
@@ -0,0 +1,318 @@
+     * like `blockIdx.x` minus number of DP workgroups.
+     */
+    CK_TILE_DEVICE void
+    get_iter_boundaries(index_t& iter, index_t& iter_end, index_t cta_idx) const noexcept;
```
Can we call iter -> iter_start?

ecamartins replied:
Updated in [a4689cb](https://github.com/ROCm/composable_kernel/pull/3018/commits/a4689cb90608d8985e2df8ee3cf5bc889a9f137f). I also updated the test file to follow this naming as well.

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm_tile_partitioner.hpp:
```diff
@@ -0,0 +1,318 @@
+     * iteration by the function.
+     * @param tile_idx       The 1D C tensor tile index for the workgroup.
+     */
+    CK_TILE_DEVICE void get_tile_boundaries(index_t& tile_iter,
```
Same here can we call tile_iter -> tile_iter_start?

ecamartins replied:
Updated in [a4689cb](https://github.com/ROCm/composable_kernel/pull/3018/commits/a4689cb90608d8985e2df8ee3cf5bc889a9f137f). I also updated the test file to follow this naming as well.

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm_tile_partitioner.hpp:
```diff
@@ -0,0 +1,318 @@
+     * @return index_t  The local starting iteration. The value is in range [0, `iters_per_tile_`).
+     * @note  Assumes `iter` >= `tile_iter`.
+     */
+    CK_TILE_DEVICE static index_t get_local_iter(index_t iter, index_t tile_iter) noexcept;
```
Can we add documentation for tile_iter and what that means?

ecamartins replied:
Thanks! This was missed. I've added it in [a4689cb](https://github.com/ROCm/composable_kernel/pull/3018/commits/a4689cb90608d8985e2df8ee3cf5bc889a9f137f).

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm_tile_partitioner.hpp:
```diff
@@ -0,0 +1,318 @@
+     * @param acc_element_bytes  The number of bytes for the accumulator data type used in the GEMM.
+     * @return index_t           The number of bytes needed for the partials and flags buffers.
+     */
+    CK_TILE_HOST index_t get_workspace_size(index_t acc_element_bytes) const noexcept;
```
I think I'd prefer to move the documentation from member variables to the getter functions

ecamartins replied:
Sounds good. For each member variable that has a getter, I've moved the documentation to the getter. See [a4689cb](https://github.com/ROCm/composable_kernel/pull/3018/commits/a4689cb90608d8985e2df8ee3cf5bc889a9f137f).

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm_tile_partitioner_impl.hpp:
```diff
@@ -0,0 +1,301 @@
+// Copyright © Advanced Micro Devices, Inc., or its affiliates.
+// SPDX-License-Identifier:  MIT
+
```
Can include the parent file here in case the _impl file gets included from somewhere else.

ecamartins replied:
Sounds good! Added in [a4689cb](https://github.com/ROCm/composable_kernel/pull/3018/commits/a4689cb90608d8985e2df8ee3cf5bc889a9f137f).

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm_tile_partitioner_impl.hpp:
```diff
@@ -0,0 +1,301 @@
+StreamKTilePartitionerBase<BlockGemmShapeType, ReductionStrategy>::get_iter_boundaries(
+    index_t& iter, index_t& iter_end, index_t cta_idx) const noexcept
+{
+    index_t extra_iters__before_me = ck_tile::min(cta_idx, extra_iters_);
```
underscore is doubled

ecamartins replied:
Thanks! Fixed in [a4689cb](https://github.com/ROCm/composable_kernel/pull/3018/commits/a4689cb90608d8985e2df8ee3cf5bc889a9f137f).

cgmillette commented on test/ck_tile/gemm_streamk/test_streamk_tile_partitioner_common.hpp:
```diff
@@ -0,0 +1,341 @@
+
+template <typename TilePartitioner>
+struct KernelWrapperSpecialized<TilePartitioner, StreamKTilePartitionerBaseMethodId::GET_LOCAL_ITER>
+    : public KernelWrapper<>
```
Forward the TilePartitioner type to the KernelWrapper?

ecamartins replied:
Not in this case. Since `get_local_iter` is static, we only need to inherit from the `KernelWrapper<>` which uses a default template parameter for the tile partitioner. This allows us to use the same parent class for testing device functions that are static and non-static.

ecamartins committed:
Style updates and cleanup

The following changes were made
- Renamed iter to iter_start
- Renamed tile_iter to tile_iter_start
- Moved documentation from member variables to getters
- Removed double underscore from extra_iters_before_me variable
- Defined parent header in impl file
- Removed unused inlcudes

cgmillette reviewed (dismissed):
LGTM!

cgmillette reviewed (approved):
lgtm! nice job
